[WIP] Moving Prompt data loading to RolloutAgent

rithwik-db opened the pull request:
Moved the prompt data loading to RolloutAgent (still WIP)

bowenyang008 commented on test_single_controller_ppo.py:
```diff
@@ -506,7 +614,13 @@ def train(self):
+            # Update the batch rollouts for all actors in the group which, at `iteration_start`, will use this latest batch_rollouts to create the
+            # dataloader for the next iteration.
+            # TODO: See if there's a way for the TrainActorGroup to handle this (maybe with ExperienceBuffer.get instead of pop?) 
+            self.train_actor.collective_methods.execute(partial(self.train_actor.update_batch_rollouts, latest_iter_data=latest_iter_data))
```
yeah, we may want to abstract it to:
train_actor.fetch() and under the hood it is experience_buffer.get(train_actor) then in the get you can just implement `collective_methods.execute(partial(self.train_actor.update_batch_rollouts, latest_iter_data=latest_iter_data))`, we will update this later

bowenyang008 replied:
and each trainer should receive its own data, now all of them received the same data

bowenyang008 commented on test_single_controller_ppo.py:
```diff
@@ -443,47 +403,194 @@ def put(self, struct: dict[str, Any]):
             sequences = vllm_generate(
-                vllm_engines=inference_server.engines,
-                batch=batch,
+                vllm_engines=self.inference_server.engines,
+                batch=iter_data,
                 max_gen_len=max_gen_len,
                 generation_kwargs=generation_kwargs,
-                tokenizer=actor.tokenizer,  # type: ignore
+                tokenizer=self.prompt_data_handler.tokenizer,  # type: ignore
                 vllm_generate_function='generate',
             )
```
we shouldn't use this anymore. and as a first step, let's implement below:
```
sequences = ray.get([engine.generate.remote(iter_data[i:i+1]) for i in range(num_vllm_engines)])
```

rithwik-db replied:
Curious on why we shouldn't use `vllm_generate` anymore?

bowenyang008 commented on test_single_controller_ppo.py:
```diff
@@ -443,47 +403,194 @@ def put(self, struct: dict[str, Any]):
+                    if key in [
+                        'prompt_len',
+                        'verified_answer',
+                        'prompt_id',
+                        'vstar',
+                        'messages',
+                    ]:
+                        curr_values.append(batch[key])
+                        continue
+
+                    bs, seq_len = batch[key].shape
+
+                    if key == 'prompt':
+                        padding_key = self.pad_token_idx
+                        if (batch[key][:, -1] == padding_key).any():
+                            raise ValueError(
+                                'The last token in the prompt should not be the pad token. Please double '
+                                +
+                                'check the dataloader and prompt and dataloader.',
+                            )
+                    elif key == 'prompt_attention_mask':
+                        padding_key = False
+
+                    # Compute the required padding and concatenate with the batch tensor
+                    pad = torch.ones(
+                        (bs, max_len - seq_len),
+                        dtype=batch[key].dtype,
+                    ) * padding_key  # type: ignore
+                    curr_values.append(torch.cat([pad, batch[key]], dim=-1))
+
+            # For tensor fields, use torch.cat to combine the values; for string fields, just use the list
+            if isinstance(curr_values[0], torch.Tensor):
+                ret_batch[key] = torch.cat(curr_values)
+            else:
+                if key in ['verified_answer', 'vstar']:
+                    ret_batch[key] = list(flatten(curr_values))
+                else:
+                    ret_batch[key] = curr_values
+
+        return ret_batch
```
let's just make a util to share with the Callback instead of copy pasting it

abaheti95 replied:
We don't want to break the callback during the hackathon. We can deprecate it once the single controller is fully checked into main.

abaheti95 commented on test_single_controller_ppo.py:
```diff
@@ -443,47 +403,194 @@ def put(self, struct: dict[str, Any]):
+        # Since this is the prompt dataloader, we only need to return the number of prompts per batch.
+        # Therefore, when vllm_generate is called, it will return num_prompts_per_batch * generations_per_prompt sequences
+        # which will equal the total number of generations for the next iteration
+        foundry_dataspec = build_dataloader(
```
If I understand correctly, this dataloader will still be distributed across the train actor GPUs, right?
This was a limitation in a callback design where we couldn't have `generations_per_prompt` > `num_batches_per_update`. Ideally, it would be preferrable to have a single instance of dataloader in the controller node such that we can tweak the number of prompts (for example, one should say 512 generations_per_prompt, 64 global_train_batch_size... that should only pull one prompt from `_get_single_batch_prompts` and we can decouple this function with `num_batches_per_update`

rithwik-db replied:
Also as a sidenote, can you remind me what are the key params that are set by the user for the prompt dataloader?
- `global_train_batch_size` -> `device_train_batch_size`
- `max_gen_len`
- `max_seq_len`
- `generations_per_prompt`
- `num_batches_per_update`

Is the above everything?

rithwik-db commented:
Too lazy to salvage this PR and clean up all changes, so moved to a separate PR: https://github.com/databricks/compose-rl/pull/125 (which addressed all comments on this PR)
